zilliqa/src/p2p_node.rs

```diff
@@ -306,27 +310,39 @@ impl P2pNode {
                                     let to = self.peer_id;
                                     let (shard_id, external_message) = request;
                                     debug!(%source, %to, %external_message, "message received");
                                     let topic = Self::shard_id_to_topic(shard_id);
-                                    self.forward_external_message_to_node(&topic.hash(), source, external_message)?;
+                                    self.forward_external_message_to_node(&topic.hash(), source, Ok(external_message))?;
                                     let _ = self.swarm.behaviour_mut().request_response.send_response(channel, (shard_id, ExternalMessage::RequestResponse));
                                 }
-                                request_response::Message::Response {..} => {}
+                                request_response::Message::Response { request_id, .. } => {
+                                    self.pending_requests.remove(&request_id);
+                                }
                             }
                         }
-                        SwarmEvent::Behaviour(BehaviourEvent::RequestResponse(request_response::Event::OutboundFailure { error: OutboundFailure::DialFailure, .. })) => {
-                            // We failed to send a message to a peer. The likely reason is that we don't know their
-                            // address. Someone else in the network must know it, because we learnt their peer ID.
-                            // Therefore, we can attempt to learn their address by triggering a Kademlia bootstrap.
-                            let _ = self.swarm.behaviour_mut().kademlia.bootstrap();
+                        SwarmEvent::Behaviour(BehaviourEvent::RequestResponse(request_response::Event::OutboundFailure { peer, request_id, error })) => {
+                            if let OutboundFailure::DialFailure = error {
+                                // We failed to send a message to a peer. The likely reason is that we don't know their
+                                // address. Someone else in the network must know it, because we learnt their peer ID.
+                                // Therefore, we can attempt to learn their address by triggering a Kademlia bootstrap.
+                                let _ = self.swarm.behaviour_mut().kademlia.bootstrap();
+                            }
+
+
+                            if let Some((shard_id, request_id)) = self.pending_requests.remove(&request_id) {
+                                let error = OutgoingMessageFailure { peer, request_id, error };
+                                self.forward_external_message_to_node(&Self::shard_id_to_topic(shard_id).hash(), peer, Err(error))?;
+                            } else {
+                                return Err(anyhow!("request without id failed"));
+                            }
```

Comment:
This deals with the case where you can't send a block request to a peer; how about if it never replies?

Reply:
Sorry should have clarified this in the PR description. I have a future plan to alter how we request blocks which I think will fix this.

`libp2p_request_response` requires that every request has a response and will report a request failure if a response isn't received within a configurable amount of time. However, currently we aren't utilising this because of the way we've structured our messages:

**Current state**
A -> B: `libp2p.send_request(BlockRequest)`
B -> A: `libp2p.send_response(Ack)`
B -> A: `libp2p.send_request(BlockResponse)`
B -> A: `libp2p.send_response(Ack)`

**Desired state**
A -> B: `libp2p.send_request(BlockRequest)`
B -> A: `libp2p.send_response(BlockResponse)`

This will mean a node's failure to respond to a `BlockRequest` will result in an `OutgoingMessageFailure` (of type `OutboundFailure::Timeout`) and this error handling code will be called.

If you agree with this plan, I'll raise an issue to do this before merging this PR :)

Reply:
I agree with this plan, but check (in a separate issue, I guess?) that malicious nodes can't crash the network by causing a lot of messages to be sent to them and then vanishing, repeatedly, causing the calling node to run out of memory to store pending request records. It's one of those failure modes that is either trivial or impossible, so worth checking - if you're lucky, there is a max # pending messages in libp2p and you can set it high enough to make that situation impossible.